File: pyround1/538.把二叉搜索树转换为累加树.py
#
# @lc app=leetcode.cn id=538 lang=python
#
# [538] 把二叉搜索树转换为累加树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def convertBST(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        self.num = 0
        def depth(root):
            if root is None:
                return 
            else:
                depth(root.right)
                self.num = self.num + root.val
                root.val = self.num
                depth(root.left)
                return root
        return depth(root)
# @lc code=end

File: 538.把二叉搜索树转换为累加树.cpp
/*
 * @lc app=leetcode.cn id=538 lang=cpp
 *
 * [538] 把二叉搜索树转换为累加树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int cur_val = 0;
    TreeNode* convertBST(TreeNode* root) {
        helper(root);
        return root;
    }
    int helper(TreeNode* root)
    {
        if(!root)
            return 0;
        if(!root->left and !root->right)
        {
            cur_val += root->val;
            root->val = cur_val;
            return cur_val;
        }
        if(root->right){
            helper(root->right);
        }
        root->val += cur_val;
        cur_val = root->val;
        if(root->left){
            helper(root->left);
        }
        return 0;
    }
};
// @lc code=end

